tools/calyx-writer/lib.rs

```diff
@@ -0,0 +1,1077 @@
+                writeln!(f, "par {{")?;
+                f.push();
+                for node in body {
+                    node.writeln(f)?;
+                }
+                f.pop();
+                write!(f, "}}")?;
+            }
+            ControlValue::While(cond, group, body) => {
+                writeln!(
+                    f,
+                    "while {}{} {{",
+                    cond,
+                    if let Some(group) = group {
+                        format!(" with {}", group.borrow().name)
+                    } else {
+                        "".into()
+                    }
+                )?;
+                f.push();
+                for node in body {
+                    node.writeln(f)?;
+                }
+                f.pop();
+                write!(f, "}}")?;
+            }
+            ControlValue::If(cond, group, body_true, body_false) => {
+                writeln!(
+                    f,
+                    "if {}{} {{",
+                    cond,
+                    if let Some(group) = group {
+                        format!(" with {}", group.borrow().name)
+                    } else {
+                        "".into()
+                    }
+                )?;
+                f.push();
+                for node in body_true {
+                    node.writeln(f)?;
+                }
+                f.pop();
+                writeln!(f, "}} else {{")?;
+                f.push();
+                for node in body_false {
+                    node.writeln(f)?;
+                }
+                f.pop();
+                write!(f, "}}")?;
+            }
+        }
+        Ok(())
+    }
+}
+
+/// See `calyx_ir::Component`.
+pub struct Component {
+    attributes: Attributes,
+    is_comb: bool,
+    name: String,
+    inputs: Vec<Port>,
+    outputs: Vec<Port>,
+    cells: Vec<RRC<Cell>>,
+    groups: Vec<RRC<Group>>,
+    control: Control,
+    continuous_assignments: Vec<Assignment>,
+}
+
+impl Component {
+    fn new<S: ToString>(name: S, is_comb: bool) -> Self {
+        Self {
+            attributes: vec![],
+            is_comb,
+            name: name.to_string(),
+            inputs: vec![
+                Port::new_in_comp("go", 1)
+                    .with_attribute(Attribute::bool("go")),
+                Port::new_in_comp("clk", 1)
+                    .with_attribute(Attribute::bool("clk")),
+                Port::new_in_comp("reset", 1)
+                    .with_attribute(Attribute::bool("reset")),
+            ],
+            outputs: vec![Port::new_in_comp("done", 1)
+                .with_attribute(Attribute::bool("done"))],
+            cells: vec![],
+            groups: vec![],
+            control: Control::empty(),
+            continuous_assignments: vec![],
+        }
+    }
+
+    /// Sets whether the component is combinational or not.
+    pub fn set_comb(&mut self, is_comb: bool) {
+        self.is_comb = is_comb;
+    }
```

Comment:
Slight issue with having `comb` components and normal components represented by
the same construct: `comb` components don't have go/done signals which you
instantiate in the constructor